File: VB/DockLayoutSelectExample/Behavior/DockLayoutSelectBehavior.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Mvvm.UI.Interactivity
Imports DevExpress.Xpf.Bars
Imports DevExpress.Xpf.Docking
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data
Imports System.Windows.Input
Imports System.Windows.Markup

Namespace DockLayoutSelectExample
	Friend Class DockLayoutSelectBehavior
		Inherits Behavior(Of UIElement)
		Public Shared ReadOnly DockLayoutSourceProperty As DependencyProperty = DependencyProperty.Register("DockLayoutSource", GetType(DockLayoutManager), GetType(DockLayoutSelectBehavior), New PropertyMetadata(Nothing))

		Public Property DockLayoutSource() As DockLayoutManager
			Get
				Return CType(GetValue(DockLayoutSourceProperty), DockLayoutManager)
			End Get
			Set(ByVal value As DockLayoutManager)
				SetValue(DockLayoutSourceProperty, value)
			End Set
		End Property

		Private SubItem As BarSubItem
		Private BarManager As BarManager

		Protected Overrides Sub OnAttached()
			MyBase.OnAttached()
			BarManager = TryCast(Me.AssociatedObject, BarManager)
			AddHandler BarManager.Loaded, AddressOf barManager_Loaded
		End Sub

		Private Sub barManager_Loaded(ByVal sender As Object, ByVal e As RoutedEventArgs)
			DockLayoutSource = TryCast(BarManager.Child, DockLayoutManager)
			AddHandler DockLayoutSource.Loaded, AddressOf Manager_Loaded
		End Sub

		Private Sub Manager_Loaded(ByVal sender As Object, ByVal e As RoutedEventArgs)
			CreateBar(BarManager)
		End Sub

		Public Sub CreateBar(ByVal bar As BarManager)
			SubItem = New BarSubItem()
			Dim subBar As New Bar()

			SubItem.Content = "Select focused panel"

			CreateSubItemLinks()

			bar.Items.Add(SubItem)
			subBar.ItemLinks.Add(SubItem)
			bar.Bars.Add(subBar)

			AddHandler SubItem.Popup, AddressOf subItem_Popup
		End Sub

		Private Sub checkItem_CheckedChanged(ByVal sender As Object, ByVal e As ItemClickEventArgs)
			Dim barItem = TryCast(sender, BarCheckItem)
			Dim panel = TryCast(barItem.Tag, BaseLayoutItem)
			DockLayoutSource.BringToFront(panel)
			DockLayoutSource.Activate(panel)
		End Sub

		Private Sub subItem_Popup(ByVal sender As Object, ByVal e As EventArgs)
			CreateSubItemLinks()
		End Sub

		Private Sub CreateSubItemLinks()
            Dim list = DockLayoutSource.GetItems().Where(Function(n) (n.ItemType = DevExpress.Xpf.Layout.Core.LayoutItemType.Document OrElse n.ItemType = DevExpress.Xpf.Layout.Core.LayoutItemType.Panel)).ToList()

			For Each linkItem As BarItemLinkBase In SubItem.ItemLinks
				Dim baseItem = TryCast(linkItem, BarCheckItemLink)
				If baseItem IsNot Nothing Then
					Dim checkItem = TryCast(baseItem.Item, BarCheckItem)
					RemoveHandler checkItem.CheckedChanged, AddressOf checkItem_CheckedChanged
				End If
			Next linkItem

			SubItem.ItemLinks.Clear()

			For Each item As BaseLayoutItem In list
				Dim checkItem = New BarCheckItem()
				checkItem.Content = item.Caption.ToString()
				checkItem.Tag = item

				If DockLayoutSource.ActiveLayoutItem Is item OrElse DockLayoutSource.ActiveDockItem Is item Then
					checkItem.IsChecked = True
				Else
					checkItem.IsChecked = False
				End If

				SubItem.ItemLinks.Add(checkItem)

				AddHandler checkItem.CheckedChanged, AddressOf checkItem_CheckedChanged
			Next item
		End Sub

		Protected Overrides Sub OnDetaching()
			MyBase.OnDetaching()
		End Sub
	End Class
End Namespace

File: CS/DockLayoutSelectExample/Behavior/DockLayoutSelectBehavior.cs
using DevExpress.Mvvm.UI.Interactivity;
using DevExpress.Xpf.Bars;
using DevExpress.Xpf.Docking;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Data;
using System.Windows.Input;
using System.Windows.Markup;

namespace DockLayoutSelectExample
{
    class DockLayoutSelectBehavior : Behavior<UIElement>
    {
        public static readonly DependencyProperty DockLayoutSourceProperty = DependencyProperty.Register("DockLayoutSource", typeof(DockLayoutManager), typeof(DockLayoutSelectBehavior), new PropertyMetadata(null));

        public DockLayoutManager DockLayoutSource
        {
            get { return (DockLayoutManager)GetValue(DockLayoutSourceProperty); }
            set { SetValue(DockLayoutSourceProperty, value); }
        }

        BarSubItem SubItem;
        BarManager BarManager;

        protected override void OnAttached()
        {
            base.OnAttached();
            BarManager = this.AssociatedObject as BarManager;
            BarManager.Loaded += barManager_Loaded;
        }

        void barManager_Loaded(object sender, RoutedEventArgs e)
        {
            DockLayoutSource = BarManager.Child as DockLayoutManager;
            DockLayoutSource.Loaded += Manager_Loaded;
        }

        void Manager_Loaded(object sender, RoutedEventArgs e)
        {
            CreateBar(BarManager);
        }

        public void CreateBar(BarManager bar)
        {
            SubItem = new BarSubItem();
            Bar subBar = new Bar();

            SubItem.Content = "Select focused panel";

            CreateSubItemLinks();

            bar.Items.Add(SubItem);
            subBar.ItemLinks.Add(SubItem);
            bar.Bars.Add(subBar);

            SubItem.Popup += subItem_Popup;
        }

        void checkItem_CheckedChanged(object sender, ItemClickEventArgs e)
        {
            var barItem = sender as BarCheckItem;
            var panel = barItem.Tag as BaseLayoutItem;
            DockLayoutSource.BringToFront(panel);
            DockLayoutSource.Activate(panel);
        }

        void subItem_Popup(object sender, EventArgs e)
        {
            CreateSubItemLinks();
        }

        void CreateSubItemLinks()
        {
            var list = DockLayoutSource.GetItems().Where(n => (n.ItemType == DevExpress.Xpf.Layout.Core.LayoutItemType.Document || n.ItemType == DevExpress.Xpf.Layout.Core.LayoutItemType.Panel)).ToList<BaseLayoutItem>();

            foreach(BarItemLinkBase linkItem in SubItem.ItemLinks)
            {
                var baseItem = linkItem as BarCheckItemLink;
                if (baseItem != null)
                {
                    var checkItem = baseItem.Item as BarCheckItem;
                    checkItem.CheckedChanged -= checkItem_CheckedChanged;
                }
            }

            SubItem.ItemLinks.Clear();

            foreach (BaseLayoutItem item in list)
            {
                var checkItem = new BarCheckItem();
                checkItem.Content = item.Caption.ToString();
                checkItem.Tag = item;

                if (DockLayoutSource.ActiveLayoutItem == item || DockLayoutSource.ActiveDockItem == item)
                {
                    checkItem.IsChecked = true;
                }
                else
                {
                    checkItem.IsChecked = false;
                }

                SubItem.ItemLinks.Add(checkItem);

                checkItem.CheckedChanged += checkItem_CheckedChanged;
            }
        }

        protected override void OnDetaching()
        {
            base.OnDetaching();
        }
    }
}
